src/Elasticsearch/Metadata/ElasticsearchDocument.php

```diff
@@ -0,0 +1,35 @@
+<?php
+
+/*
+ * This file is part of the API Platform project.
+ *
+ * (c) Kévin Dunglas <dunglas@gmail.com>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+declare(strict_types=1);
+
+namespace ApiPlatform\Elasticsearch\Metadata;
+
+use ApiPlatform\Metadata\PersistenceMeansInterface;
+
+final class ElasticsearchDocument implements PersistenceMeansInterface
+{
+    public function __construct(
+        public readonly ?string $index = null,
+        public readonly ?string $type = null,
+    ) {
+    }
+
+    public function withIndex(?string $index): self
+    {
+        return new self($index, $this->type);
```

Comment:
Check how we do in the metadata:

```
    public function getOperations(): ?Operations
    {
        return $this->operations;
    }

    public function withOperations(Operations $operations): self
    {
        $self = clone $this;
        $self->operations = $operations;

        return $self;
    }
``` 

It's easier to evolve if we add fields to the constructor.